aten/src/ATen/TensorImpl.cpp

```diff
@@ -89,6 +91,34 @@ IntList TensorImpl::strides() const {
   return strides_;
 }
 
+int64_t TensorImpl::numel() const {
+#ifdef DEBUG
+  int64_t n = 1;
+  for (auto s : sizes()) {
+    n *= s;
+  }
+  AT_ASSERT(n == numel_);
```

Comment:
This code is a bit copy-pasted, with `refresh_numel`. Another way to restructure this code is as follows: create a non-mutating function `compute_numel` which computes the numel directly from sizes. Implement `refresh_numel` as `numel_ = compute_numel()`. Now your ASSERT can be `AT_ASSERT(compute_numel() == numel_)`

And you should do this for `is_contiguous()` too.